src/zjs_callbacks.c

```diff
@@ -387,64 +396,113 @@ void print_callbacks(void)
+                if (cb_map[id]->js->post) {
+                    cb_map[id]->js->post(cb_map[id]->js->handle, &ret_val);
+                }
+                if (cb_map[id]->js->once) {
+                    zjs_remove_callback(id);
+                }
+            } else {
+                // Function list callback
+                int i;
+                jerry_value_t ret_val;
 
-            for (j = 0; j < cb_map[i]->js->num_funcs; ++j) {
-                ret_val = jerry_call_function(cb_map[i]->js->func_list[j], cb_map[i]->js->this, args, argc);
-                if (jerry_value_has_error_flag(ret_val)) {
-                    DBG_PRINT("callback %ld returned an error for function[%i]\n", i, j);
+                for (i = 0; i < cb_map[id]->js->num_funcs; ++i) {
+                    ret_val = jerry_call_function(cb_map[id]->js->func_list[i], cb_map[id]->js->this, data, sz);
+                    if (jerry_value_has_error_flag(ret_val)) {
+                        DBG_PRINT("callback %ld returned an error for function[%i]\n", id, i);
+                    }
+                }
+                if (cb_map[id]->js->post) {
+                    cb_map[id]->js->post(cb_map[id]->js->handle, &ret_val);
                 }
             }
-            if (cb_map[i]->js->post) {
-                cb_map[i]->js->post(cb_map[i]->js->handle, &ret_val);
-            }
+        } else if (cb_map[id]->type == CALLBACK_TYPE_C && cb_map[id]->c->function) {
+            cb_map[id]->c->function(cb_map[id]->c->handle, data);
         }
-    } else if (cb_map[i]->type == CALLBACK_TYPE_C && cb_map[i]->c->function) {
-        DBG_PRINT("calling callback id %ld\n", cb_map[i]->c->id);
-        cb_map[i]->c->function(cb_map[i]->c->handle);
+    } else {
+        ERR_PRINT("callback does not exist: %ld\n", id);
     }
 }
 
 void zjs_service_callbacks(void)
 {
-    int i;
-    for (i = 0; i < cb_size; i++) {
-        if (cb_map[i] && cb_map[i]->signal) {
-            cb_map[i]->signal = 0;
-            zjs_call_callback(i);
+    if (ring_buf_initialized) {
+#ifdef ZJS_PRINT_CALLBACK_STATS
+        uint8_t header_printed = 0;
+        uint32_t num_callbacks = 0;
+#endif
+        uint16_t count = 0;
+        while (count++ < ZJS_MAX_CB_LOOP_ITERATION) {
+            int ret;
+            uint16_t id;
+            uint8_t value;
+            uint8_t size = 0;
+            // setting size = 0 will check if there is an item in the ring buffer
+            ret = zjs_port_ring_buf_get(&ring_buffer,
+                                        &id,
+                                        &value,
+                                        NULL,
+                                        &size);
+            if (ret == -EMSGSIZE || ret == 0) {
+                // item in ring buffer with size > 0, has args
+                // pull from ring buffer
+                int ret1;
+                uint16_t id1;
+                uint8_t value1;
```

Comment:
You're only overwriting these three in the -ENOMSGSIZE case so you don't actually need new variables, you can overwrite the old ones.